File: python/lib/lineardoc/builder.py
"""
Builder - A document builder for creating linear documents.
"""

from . import utils
from .doc import Doc
from .text_block import text_block
from .text_chunk import TextChunk


class Builder:
    """A document builder."""

    def __init__(self, parent=None, wrapper_tag=None):
        """
        Initialize a Builder.

        Args:
            parent: Parent document builder
            wrapper_tag: Tag that wraps document (if there is a parent)
        """
        self.block_tags = []
        # Stack of annotation tags
        self.inline_annotation_tags = []
        # The height of the annotation tags that have been used, minus one
        self.inline_annotation_tags_used = 0
        self.doc = Doc(wrapper_tag)
        self.text_chunks = []
        self.is_block_segmentable = True
        self.parent = parent

    def create_child_builder(self, wrapper_tag):
        """
        Create a child builder.

        Args:
            wrapper_tag: Wrapper tag for child

        Returns:
            New Builder instance
        """
        return Builder(self, wrapper_tag)

    def push_block_tag(self, tag):
        """Push a block tag."""
        self.finish_text_block()
        self.block_tags.append(tag)
        if self.is_ignored_tag(tag):
            return
        if tag["name"] == "figure":
            tag["attributes"]["rel"] = "cx:Figure"
        self.doc.add_item("open", tag)

    def is_section(self, tag):
        """Check if tag is a section."""
        return tag["name"] == "section" and tag.get("attributes", {}).get("data-mw-section-id")

    def is_ignored_tag(self, tag):
        """Check if tag should be ignored."""
        return self.is_section(tag) or self.is_category(tag)

    def is_category(self, tag):
        """Check if tag is a category."""
        # content can be a Doc instance, not just a tag
        if not isinstance(tag, dict):
            return False
        rel = tag.get("attributes", {}).get("rel", "")
        return (
            tag["name"] == "link"
            and f" {rel} ".find(" mw:PageProp/Category ") != -1
            and not tag.get("attributes", {}).get("about")
        )

    def pop_block_tag(self, tag_name):
        """
        Pop a block tag.

        Args:
            tag_name: Name of tag to pop

        Returns:
            The popped tag
        """
        if not self.block_tags:
            tag = None
        else:
            tag = self.block_tags.pop()

        if not tag or tag["name"] != tag_name:
            raise Exception(f"Mismatched block tags: open={tag['name'] if tag else None}, close={tag_name}")

        self.finish_text_block()

        if not self.is_ignored_tag(tag):
            self.doc.add_item("close", tag)

        return tag

    def push_inline_annotation_tag(self, tag):
        """Push an inline annotation tag."""
        self.inline_annotation_tags.append(tag)

    def pop_inline_annotation_tag(self, tag_name):
        """Pop an inline annotation tag."""
        if not self.inline_annotation_tags:
            tag = None
        else:
            tag = self.inline_annotation_tags.pop()

        if self.inline_annotation_tags_used == len(self.inline_annotation_tags):
            self.inline_annotation_tags_used -= 1

        if not tag or tag["name"] != tag_name:
            raise Exception(f"Mismatched inline tags: open={tag['name'] if tag else None}, close={tag_name}")

        if not tag.get("attributes"):
            # Skip tags which have no attributes
            return

        # Check for empty/whitespace-only data tags
        replace = True
        whitespace = []
        i = len(self.text_chunks) - 1
        while i >= 0:
            text_chunk = self.text_chunks[i]
            chunk_tag = text_chunk.tags[-1] if text_chunk.tags else None
            if not chunk_tag:
                break
            if text_chunk.text.strip() or text_chunk.inline_content or chunk_tag is not tag:
                # text_chunk has non whitespace content
                replace = False
                break
            whitespace.append(text_chunk.text)
            i -= 1

        # Allow empty external links and references
        if replace and (utils.is_reference(tag) or utils.is_external_link(tag) or utils.is_transclusion(tag)):
            # truncate list and add data span as new sub-Doc
            self.text_chunks = self.text_chunks[: i + 1]
            whitespace.reverse()
            self.add_inline_content(
                Doc()
                .add_item("open", tag)
                .add_item("textblock", text_block([TextChunk("".join(whitespace), [])]))
                .add_item("close", tag)
            )

    def add_text_chunk(self, text, can_segment):
        """
        Add a text chunk.

        Args:
            text: Text content
            can_segment: Whether this can be segmented
        """
        self.text_chunks.append(TextChunk(text, self.inline_annotation_tags[:]))
        self.inline_annotation_tags_used = len(self.inline_annotation_tags)
        # Inside a textblock, if a textchunk becomes segmentable
        self.is_block_segmentable = can_segment

    def add_inline_content(self, content, can_segment=True) -> None:
        """
        Add content that doesn't need linearizing, to appear inline.

        Args:
            content: Sub-document or empty SAX tag
            can_segment: Whether this can be segmented
        """
        # If the content is a category tag, capture it separately
        if self.is_category(content):
            self.doc.categories.append(content)
            return

        self.text_chunks.append(TextChunk("", self.inline_annotation_tags[:], content))
        self.inline_annotation_tags_used = len(self.inline_annotation_tags)
        if not can_segment:
            self.is_block_segmentable = False

    def finish_text_block(self):
        """Finish the current text block."""
        if len(self.text_chunks) == 0:
            return

        whitespace = []
        whitespace_only = True

        for text_chunk in self.text_chunks:
            if text_chunk.inline_content or text_chunk.text.strip():
                whitespace_only = False
                break
            else:
                whitespace.append(text_chunk.text)

        if whitespace_only:
            self.doc.add_item("blockspace", "".join(whitespace))
        else:
            self.doc.add_item("textblock", text_block(self.text_chunks, self.is_block_segmentable))

        self.text_chunks = []
        self.is_block_segmentable = True

File: cxsever/www/js/lib/lineardoc/Builder.js
import Doc from './Doc.js';
import { isExternalLink, is_reference, isTransclusion } from './utils.js';
import text_block from './text_block.js';
import text_chunk from './text_chunk.js';

/**
 * A document builder
 *
 * @class
 */
class Builder {
	/**
	 * @param {Builder} [parent] Parent document builder
	 * @param {Object} [wrapper_tag] tag that wraps document (if there is a parent)
	 */
	constructor(parent, wrapper_tag) {
		this.BLOCK_TAGS = [];
		// Stack of annotation tags
		this.inlineAnnotationTags = [];
		// The height of the annotation tags that have been used, minus one
		this.inlineAnnotationTagsUsed = 0;
		this.doc = new Doc(wrapper_tag || null);
		this.text_chunks = [];
		this.isBlockSegmentable = true;
		this.parent = parent || null;
	}

	createChildBuilder(wrapper_tag) {
		return new Builder(this, wrapper_tag);
	}

	pushBlockTag(tag) {
		this.finishTextBlock();
		this.BLOCK_TAGS.push(tag);
		if (this.isIgnoredTag(tag)) {
			return;
		}
		if (tag.name === 'figure') {
			tag.attributes.rel = 'cx:Figure';
		}
		this.doc.add_item('open', tag);
	}

	isSection(tag) {
		return tag.name === 'section' && tag.attributes['data-mw-section-id'];
	}

	isIgnoredTag(tag) {
		return this.isSection(tag) || this.isCategory(tag);
	}

	isCategory(tag) {
		return tag.name === 'link' && tag.attributes.rel &&
			// We add the spaces before and after to ensure matching on the "word" mw:PageProp/Category
			// without additional content. This is technically not necessary (we don't generate
			// mw:PageProp/Category/SomethingElse) nor entirely correct (attributes values could be separated by other
			// characters than 0x20), but provides a bit of future-proofing.
			(' ' + tag.attributes.rel + ' ').includes(' mw:PageProp/Category ') && !tag.attributes.about;
	}

	pop_block_tag(tagName) {
		const tag = this.BLOCK_TAGS.pop();
		if (!tag || tag.name !== tagName) {
			throw new Error(
				'Mismatched block tags: open=' + (tag && tag.name) + ', close=' + tagName
			);
		}
		this.finishTextBlock();

		if (!this.isIgnoredTag(tag)) {
			this.doc.add_item('close', tag);
		}

		return tag;
	}

	pushInlineAnnotationTag(tag) {
		this.inlineAnnotationTags.push(tag);
	}

	popInlineAnnotationTag(tagName) {
		let i;
		const tag = this.inlineAnnotationTags.pop();
		if (this.inlineAnnotationTagsUsed === this.inlineAnnotationTags.length) {
			this.inlineAnnotationTagsUsed--;
		}
		if (!tag || tag.name !== tagName) {
			throw new Error(
				'Mismatched inline tags: open=' + (tag && tag.name) + ', close=' + tagName
			);
		}

		if (!Object.keys(tag.attributes).length) {
			// Skip tags which have attributes, content or both from further check to
			// see if we want to keep inline content. Else we assume that, if this tag has
			// whitespace or empty content, it is ok to remove it from resulting document.
			// But if it has attributes, we make sure that there are inline content blocks to
			// avoid them missing in resulting document.
			// See T104539
			return;
		}
		// Check for empty/whitespace-only data tags. Replace as inline content
		let replace = true;
		const whitespace = [];
		for (i = this.text_chunks.length - 1; i >= 0; i--) {
			const textChunk = this.text_chunks[i];
			const chunkTag = textChunk.tags[textChunk.tags.length - 1];
			if (!chunkTag) {
				break;
			}
			if (textChunk.text.match(/\S/) || textChunk.inline_content || chunkTag !== tag) {
				// textChunk has non whitespace content, Or it has child tags.
				replace = false;
				break;
			}
			whitespace.push(textChunk.text);
		}

		// Allow empty external links because REST API v1 can output links with
		// no link text (which then get a CSS generated content numbered reference).
		if (replace && (is_reference(tag) || isExternalLink(tag) || isTransclusion(tag))) {
			// truncate list and add data span as new sub-Doc.
			this.text_chunks.length = i + 1;
			whitespace.reverse();
			this.addInlineContent(
				new Doc()
					.add_item('open', tag)
					.add_item('textblock', new text_block(
						[new text_chunk(whitespace.join(''), [])]
					))
					.add_item('close', tag)
			);
		}
		return;
	}

	addTextChunk(text, can_segment) {
		this.text_chunks.push(new text_chunk(text, this.inlineAnnotationTags.slice()));
		this.inlineAnnotationTagsUsed = this.inlineAnnotationTags.length;
		// Inside a textblock, if a textchunk becomes segmentable, unlike inline tags,
		// the textblock becomes segmentable. See T195768
		this.isBlockSegmentable = can_segment;
	}

	/**
	 * Add content that doesn't need linearizing, to appear inline
	 *
	 * @method
	 * @param {Object} content Sub-document or empty SAX tag
	 * @param {boolean} can_segment
	 */
	addInlineContent(content, can_segment) {
		// If the content is a category tag, capture it separately and don't add to doc.
		if (this.isCategory(content)) {
			this.doc.categories.push(content);
			return;
		}
		this.text_chunks.push(new text_chunk('', this.inlineAnnotationTags.slice(), content));
		if (!can_segment) {
			this.isBlockSegmentable = false;
		}
		this.inlineAnnotationTagsUsed = this.inlineAnnotationTags.length;
	}

	finishTextBlock() {
		let whitespace = [], whitespaceOnly = true;

		if (this.text_chunks.length === 0) {
			return;
		}
		for (let i = 0, len = this.text_chunks.length; i < len; i++) {
			const textChunk = this.text_chunks[i];
			if (textChunk.inline_content || textChunk.text.match(/\S/)) {
				whitespaceOnly = false;
				whitespace = undefined;
				break;
			} else {
				whitespace.push(this.text_chunks[i].text);
			}
		}
		if (whitespaceOnly) {
			this.doc.add_item('blockspace', whitespace.join(''));
		} else {
			this.doc.add_item('textblock', new text_block(this.text_chunks, this.isBlockSegmentable));
		}
		this.text_chunks = [];
		this.isBlockSegmentable = true;
	}

}

export default Builder;
